src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -323,6 +323,38 @@ internal List<WildcardPattern> MatchAll
                 }
             }
 
+#if UNIX    // case-sensitive filesystems are predominately used only on Unix
+            if (!found)
+            {
+                // try a case-insensitive search for the module folder
+                // based on PSGet, we can assume the module file casing matches the folder
+                foreach (string path in modulePath)
```

Comment:
In line 294 the same cycle? Can we remove it and replace with the new one on all platforms?

Reply:
We still need both as the first one will find an exact match and fall through to this if it doesn't find it.  Also the above loop also handles a file path to the module which I think should be case sensitive since it's a literal file path.  The lower code only handles the module name which we can treat as case insensitive.  The code is similar, but I don't think it's worth combining which will make it more complex.

Reply:
Why we should use different logic for module folder names and module file names?

Reply:
If someone uses: `Import-module /users/steve/dev/MyModule/MyModule.psd1` then that should be treated as a case-sensitive path.  If they use `Import-module mymodule`, then that's case-insensitive as that's a module name and not a filepath.

Reply:
Ah, I meant folder and file names not paths.

Reply:
I don't see an easy way to have a single loop between the two without incurring a cost of enumerating the folders for the purpose of case-insensitive comparison in the "normal" case where either the case matches or the filesystem is case-insensitive.  If there's two modules "Foo" and "foo", and the user does `import-module foo` we want the first loop to match `foo` and not `Foo` as that is an exact match.

Reply:
We have the same problem in #4699 - can you review the idea to use here?

Reply:
In the tab completion case, I think you have to cover more cases as `ab` might match `ABc`, `abc`, or `aBcdefg`, etc...  so creating a case-insenstive match pattern makes sense.  In this case, we have a specific name (wildcard is not accepted) and just need a case-insensitive match if we haven't already matched the exact casing (which is preferred).  And we don't need to return all matches, the first one wins.

If we want just the first match to win even if there is an exact match, it would be easy to merge the two loops.  However, I think we want to get an exact match if it exists, then fallback on case-sensitive search.  For example, if the user typed `ipmo AAA` which exists in the last modulepath, but `aaa` exists in the first, I would have to enumerate all module folder names in every modulepath part every time just to make sure `AAA` wasn't found after my first case-insensitive match.

Reply:
We could have solved this situation on the other hand. Do we want _all_ the modules to be ported? If so, we must document this and exclude the case-sensitivity in this code.

Reply:
That's a good point on portable scripts being able to rely on module names being case insensitive.  It would make sense to just say module names are case insensitive everywhere and have it enforced by install-module and publish-module.